hudi-common/src/main/java/org/apache/hudi/common/table/cdc/HoodieCDCExtractor.java

```diff
@@ -0,0 +1,355 @@
+              }
+            } catch (IOException e) {
+              throw new HoodieIOException(e.getMessage());
+            }
+            return Pair.of(instant, commitMetadata);
+          }).filter(pair ->
+              maybeChangeData(pair.getRight().getOperationType())
+          ).collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
+      this.commits = result;
+    } catch (Exception e) {
+      throw new HoodieIOException("Fail to get the commit metadata for CDC");
+    }
+  }
+
+  private Boolean maybeChangeData(WriteOperationType operation) {
+    return operation == WriteOperationType.INSERT
+        || operation == WriteOperationType.UPSERT
+        || operation == WriteOperationType.DELETE
+        || operation == WriteOperationType.BULK_INSERT
+        || operation == WriteOperationType.DELETE_PARTITION
+        || operation == WriteOperationType.INSERT_OVERWRITE
+        || operation == WriteOperationType.INSERT_OVERWRITE_TABLE
+        || operation == WriteOperationType.BOOTSTRAP;
+  }
+
+  /**
+   * Parse HoodieWriteStat, judge which type the file is, and what strategy should be used to parse CDC data.
+   * Then build a [[HoodieCDCFileSplit]] object.
+   */
+  private HoodieCDCFileSplit parseWriteStat(
+      HoodieFileGroupId fileGroupId,
+      HoodieInstant instant,
+      HoodieWriteStat writeStat,
+      WriteOperationType operation) {
+    Path basePath = metaClient.getBasePathV2();
+    FileSystem fs = metaClient.getFs().getFileSystem();
+
+    HoodieCDCFileSplit cdcFileSplit;
+    if (StringUtils.isNullOrEmpty(writeStat.getCdcPath())) {
+      // no cdc log files can be used directly. we reuse the existing data file to retrieve the change data.
+      String path = writeStat.getPath();
+      if (FSUtils.isBaseFile(new Path(path))) {
+        // this is a base file
+        if (operation == WriteOperationType.DELETE && writeStat.getNumWrites() == 0L
+            && writeStat.getNumDeletes() != 0) {
+          // This is a delete operation wherein all the records in this file group are deleted
+          // and no records have been writen out a new file.
+          // So, we find the previous file that this operation delete from, and treat each of
+          // records as a deleted one.
+          HoodieBaseFile beforeBaseFile = fsView.getBaseFileOn(
+              fileGroupId.getPartitionPath(), writeStat.getPrevCommit(), fileGroupId.getFileId()
+          ).orElseThrow(() ->
+              new HoodieIOException("Can not get the previous version of the base file")
+          );
+          FileSlice beforeFileSlice = new FileSlice(fileGroupId, writeStat.getPrevCommit(), beforeBaseFile, new ArrayList<>());
+          cdcFileSplit = new HoodieCDCFileSplit(BASE_FILE_DELETE, null, Option.empty(), Option.of(beforeFileSlice));
+        } else if (writeStat.getNumUpdateWrites() == 0L && writeStat.getNumDeletes() == 0
+            && writeStat.getNumWrites() == writeStat.getNumInserts()) {
```

Comment:
@YannByron there's an issue right now where we undercount inserts (AFAIR) and so `numWrites != numUpdates + numInserts` we need to be careful with this conditionals (and address the underlying issue as well)